django/contrib/admin/filters.py

```diff
@@ -140,7 +140,7 @@ def get_facet_counts(self, pk_attname, filtered_qs):
             if lookup_qs is not None:
                 counts[f"{i}__c"] = models.Count(
                     pk_attname,
-                    filter=lookup_qs.query.where,
+                    filter=models.Q(pk__in=lookup_qs),
```

Comment:
I fear that might cause performance problems on backends that aren't especially good at planing subqueries such as MySQL as this will perform an aggregation over a subquery.

It's not a dependent subquery though and faceting is already quite expensive and unsuitable for large data sets so maybe that's not an issue? It would be great if we could avoid the subquery in most cases though instead of defaulting to it.

Reply:
@charettes Do you have suggestions on how to detect whether a full subquery is necessary? I imagine we'll need to look through the expressions in the `WhereNode` and see if there are any that are not a field on the model, but I'm not sure how to do that.

Reply:
> Do you have suggestions on how to detect whether a full subquery is necessary? 

It's quite complicated unfortunately I think the whole feature is kind of broken due to ticket-10060 which likely explains why the change list performs one `aggregate` call per list filter instead of a single one for the entirety.

I suspect that when the feature was designed it was noticed that different list filters were more likely to span different multi-valued relationships and trigger the problem described by ticket-10060 and thus it was chosen to perform one `aggregate` call per filter.

The thing is there is nothing that forces a list filter to only span a single multi-valued relationship.

e.g. an author list filter could do

```
[
    ('Has long books', Q(books__pages__gt=1000)),
    ('Has established publisher', Q(publishers__foundation__year__lte=ESTABLISHED_CUTOFF)),
]
```

In this case the count for both relationships will be [the cross-product book * publishers instead of separated ones](http://charettes.name/djangoconus2022/slides.html#21). What I'm trying to say there is that usually faceting is off mutually exclusive choices, but nothing forces members of list filters to be mutually exclusive which poses a problem when more than one multi-valued relationships is involved.

Because of the way the current `get_facet_counts` API is designed to not allow annotations that might be referenced by members of `lookup_qs.query.where` to also be returned I don't see a straightforward way of resolving the issue. We'll have to live with the subquery if we want to keep supporting this feature even if performs poorly on some backends.

If we could go back in time in think that having two distinct list filter method, one to return a queryset suitable for filtering against all values where annotations are made (e.g. `queryset(request, queryset)`, and a second one that returns `Q` object for specific values (e.g. `filter_for_value(request, queryset)`) would be more suitable.

That would allow `get_facet_queryset` to preemptively call into the method to perform the annotations to augment `filtered_qs` and then have the lookup choice loop simply call `filter_for_value` to avoid an unnecessary queryset construction and simply assigned the return value to `Count.filter`.

Reply:
Thank you for this @adzshaf I think this looks really good 👍 
This is my current set of thoughts.

In terms of documentation, the main part is in `ModelAdmin.show_facets` https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#django.contrib.admin.ModelAdmin.show_facets which also has a performance warning note. Nothing else in the API is documented, so I think we can change some things if we need to.

One thought is `SimpleListFilter` could do a count per option rather than an aggregate over a subquery if we think that performs better.

Maybe we have `get_facet_queryset` be defined on each filter and remove the `get_facet_counts`.
`get_facet_queryset` actually returns a dict after the aggregate and so maybe we want to encourage that if anyone really likes the facet filter but their `SimpleListFilter` implementation has poor performance, they can overwrite the `get_facet_queryset` method to something more performant (and probably rename `get_facet_queryset` to `get_facet_counts` because the names are confusing).

---
This also needs a note adding to the 5.0.3 release notes as the aim is to include it there 👍

Reply:
@sarahboyce Yeah, I guess we can do the counts separately if that performs better. I think we can do it by overriding `get_facet_queryset` in `SimpleListFilter` that returns a `dict` of the `lookup_qs.count()`s, leaving `get_facet_counts` unimplemented.

And yep I agree that the `get_facet_queryset` name is pretty confusing as it returns a `dict` and not a `queryset`. As for the naming, this would be my suggestion:
- Rename the current `get_facet_counts` to `get_facet_aggregates`. This reflects what the method does: collecting the aggregates to be used in `.aggregate()`.
- Rename the current `get_facet_queryset` to `get_facet_counts`. The default implementation in `FacetsMixin` would stay as-is, i.e. using `aggregate()`. This rename would make the following line read better:
  ```py
  facet_counts = self.get_facet_counts(changelist) if add_facets else None
  ```
  (The `facet_counts` variable now comes from a corresponding `get_facet_counts` method instead of `get_facet_queryset`.)

Are we OK with renaming them in a patch release, though? As you said, they're undocumented, but I just want to make sure. We can do this separately though, i.e. keep this PR to just override `get_facet_queryset()` to do a count for each option as a fix for the ticket. Then, in a separate PR, we do the rename and add a note about overriding the to-be-named `get_facet_counts` to optimize the performance. If we feel that this is too much of a change for a patch release, we can keep the rename only for 5.1.

Reply:
> Are we OK with renaming them in a patch release, though?

I think we need guidance from a fellow (cc @felixxm @nessita) 🙏

Reply:
Renaming methods is not needed to fix a regression. You can propose it separately as a cleanup in the `main` branch.

Reply:
As we already have a performance warning for using facet filters and there are ways to turn these off completely, I think let's continue with this solution and we can make incremental improvements into main (unless someone thinks of something better) 👍